### DISCUSSION TITLE
Extra compile step

### DISCUSSION THREAD

**Discussion by renatodex** [facebook/docusaurus#6106] [Q&A] [answered]
Hey folks, I don't know how to do this, so I'm posting this question here to see if anyone can help me!

I have an open source project of a Tabletop RPG System using `Docusaurus 2.0.0-alpha.65`. I know, it's not a stable version, but it's running fine.

Since it's an RPG Book, It has a lot of Skills, and each Skill is being currently defined in a JSON file, inside the `data/` folder:
https://github.com/renatodex/fabulas-e-goblins/tree/renatodex/1.6-changes

Then I import those JSON files into the markdown pages, so they can be used by a React component:
That's how they are being imported.

```
import apprentice_blast from './../../data/spells-allura/tier1/apprentice_blast.json'
import arcane_amplification from './../../data/spells-allura/tier1/arcane_amplification.json'
import create_element from './../../data/spells-allura/tier1/create_element.json'
import detect_element from './../../data/spells-allura/tier1/detect_element.json'
```

But, the process to import those Json files can be quite problematic, since I need to repeat the import for every place I want to use the JSON.
Now I'm creating a new section where I will be listing all Skills in one single searchable list, and I wanted to add an `extra compilation step` to Docusaurus server, so I **could grab every JSON file** in the `data/` folder an compress every single Spell in one single file named `spells.js`. Inside this output file, each individual JSON file of the `data/` folder would have to be converted to individual named exports of Const variables using the file name of each json. (ex: `export const double_fist = {...}`)

So basically what I wanted was to have something like this:

```js
// Example of a JSON file double_fist.json
{
  "name": "Punho Duplo",
  "magic_cost": "2x P.M",
  "duration_time": null,
  "attack_logic": null,
  "action_type": "Aprimoramento",
  "cast_distance": null,
  "range_amount": null,
  "range_type": null,
  "icon": "GiDigHole",
  "element": null,
  "caption": "Você é o mestre dos Punhos de Mago, conseguindo desferir dois ataques poderosos simultaneamente.",
  "description": "<p>Aprimora o Punhos de Mago, desferindo dois Punhos de Mago em uma única ação. Cada ataque precisa ser testado separadamente e precisa acertar o mesmo alvo.</p>",
  "tags": ["grimo", "orbe-de-allura"]
}
```

Compiled to the output file into something like this:

```js
export const DoubleFist = {
  "name": "Punho Duplo",
  "magic_cost": "2x P.M",
  "duration_time": null,
  "attack_logic": null,
  "action_type": "Aprimoramento",
  "cast_distance": null,
  "range_amount": null,
  "range_type": null,
  "icon": "GiDigHole",
  "element": null,
  "caption": "Você é o mestre dos Punhos de Mago, conseguindo desferir dois ataques poderosos simultaneamente.",
  "description": "<p>Aprimora o Punhos de Mago, desferindo dois Punhos de Mago em uma única ação. Cada ataque precisa ser testado separadamente e precisa acertar o mesmo alvo.</p>",
  "tags": ["grimo", "orbe-de-allura"]
}
```

I'm not sure what would be the best place to start with this change?
Perhaps I could create a **Docusaurus plugin**, or maybe a **Webpack override**? Or a separate bin to do the compilation in the deployment step?

What do you suggest?

Thanks for the help and for building a product so amazing!

**Comment by armano2** [comment, marked as answer]
you don't need webpack or docosaurus plugin for this,

i recommend to create spells.js file that imports all json files, with that source maps are going to point to actual files

```js
import apprentice_blast from './../../data/spells-allura/tier1/apprentice_blast.json'

export {
  apprentice_blast: apprentice_blast,
  // ....
}
```

or


```js
export const apprentice_blast = require('./../../data/spells-allura/tier1/apprentice_blast.json')
```

------------------------------

if you rly want to make this in automated way (not tested and most likely not fully working)

```js
const imports = require.context('./../../data/spells-allura/', true, /\.json$/);

const spells = imports.keys().reduce((key, acc) => {
  acc[key] = imports(key);
  return acc;
}, {});

export default spells
```

https://github.com/webpack/webpack/issues/118 https://webpack.js.org/guides/dependency-management/#requirecontext

**Reply by renatodex** [reply]
Hmmmmmmm, that's a great idea!
Thank you very much @armano2 !
